R/conditions.R

```diff
@@ -0,0 +1,254 @@
+                            what = NULL,
+                            arg = NULL,
+                            message = NULL,
+                            .recycle = FALSE,
+                            .subclass = NULL) {
+  what <- what %||% what_bad_object(arg) %||% "Vector"
+
+  if (.recycle) {
+    expected <- sprintf("1 or %s", expected_length)
+  } else {
+    expected <- as.character(expected_length)
+  }
+  actual <- length(x)
+
+  message <- message %||% sprintf(
+    "%s must have length %s, not %s",
+    what,
+    expected,
+    actual
+  )
+
+  abort(
+    message,
+    x = x,
+    expected_length = expected_length,
+    what = what,
+    arg = arg,
+    ...,
+    .subclass = c(.subclass, "purrr_error_bad_length")
+  )
+}
+
+stop_bad_element_length <- function(x,
+                                    index,
+                                    expected_length,
+                                    ...,
+                                    what = NULL,
+                                    arg = NULL,
+                                    message = NULL,
+                                    .recycle = FALSE,
+                                    .subclass = NULL) {
+  stopifnot(is_integerish(index, n = 1, finite = TRUE))
+  what <- what_bad_element(what, arg, index)
+
+  stop_bad_length(
+    x,
+    expected_length,
+    what = what,
+    arg = arg,
+    index = index,
+    ...,
+    message = message,
+    .recycle = .recycle,
+    .subclass = c(.subclass, "purrr_error_bad_element_length")
+  )
+}
+
+#' Error conditions for bad vectors
+#'
+#' @inheritParams purrr-conditions-length
+#' @param expected_ptype The expected prototype of `x`, i.e. an empty
+#'   vector of the expected type.
+#'
+#' @keywords internal
+#' @name purrr-conditions-vector
+#' @noRd
+stop_bad_vector <- function(x,
+                            expected_ptype,
+                            expected_length = NULL,
+                            ...,
+                            what = NULL,
+                            arg = NULL,
+                            message = NULL,
+                            .recycle = FALSE,
+                            .subclass = NULL) {
+  what <- what %||% what_bad_object(arg) %||% "Vector"
+
+  expected <- friendly_vector_type(expected_ptype, expected_length, .recycle)
+  actual <- friendly_vector_type(x, length(x))
+
+  stop_bad_type(
+    x,
+    expected,
+    actual = actual,
+    what = what,
+    arg = arg,
+    message = message,
+    .subclass = c(.subclass, "purrr_error_bad_vector")
+  )
+}
+
+stop_bad_element_vector <- function(x,
+                                    index,
+                                    expected_ptype,
+                                    expected_length,
+                                    ...,
+                                    what = NULL,
+                                    arg = NULL,
+                                    message = NULL,
+                                    .recycle = FALSE,
```

Comment:
I don't understand why `recycle` gets a `.` prefix

Reply:
Because it doesn't become a field of the error condition.

Reply:
Oh hmmmmmmmmm. Why doesn't it?

Reply:
It now does and the parameter is no longer prefixed with a dot.